crates/sui-core/src/authority/authority_store.rs

```diff
@@ -823,6 +824,61 @@ impl AuthorityStore {
         Ok(())
     }
 
+    pub async fn bulk_insert_live_objects(
+        perpetual_db: &AuthorityPerpetualTables,
+        live_objects: impl Iterator<Item = LiveObject>,
+        indirect_objects_threshold: usize,
+        expected_sha3_digest: &[u8; 32],
+    ) -> SuiResult<()> {
+        let mut hasher = Sha3_256::default();
+        let mut batch = perpetual_db.objects.batch();
+        for object in live_objects {
+            hasher.update(object.object_reference().2.inner());
+            match object {
+                LiveObject::Normal(object) => {
+                    let StoreObjectPair(store_object_wrapper, indirect_object) =
+                        get_store_object_pair(object.clone(), indirect_objects_threshold);
+                    batch.insert_batch(
+                        &perpetual_db.objects,
+                        std::iter::once((
+                            ObjectKey::from(object.compute_object_reference()),
+                            store_object_wrapper,
+                        )),
+                    )?;
+                    if let Some(indirect_object) = indirect_object {
+                        batch.insert_batch(
+                            &perpetual_db.indirect_move_objects,
+                            std::iter::once((indirect_object.inner().digest(), indirect_object)),
+                        )?;
+                    }
+                    if !object.is_child_object() {
+                        Self::initialize_locks(
+                            &perpetual_db.owned_object_transaction_locks,
+                            &mut batch,
+                            &[object.compute_object_reference()],
+                            false, // is_force_reset
+                        )?;
+                    }
+                }
+                LiveObject::Wrapped(object_key) => {
+                    batch.insert_batch(
+                        &perpetual_db.objects,
+                        std::iter::once::<(ObjectKey, StoreObjectWrapper)>((
+                            object_key,
+                            StoreObject::Wrapped.into(),
+                        )),
+                    )?;
+                }
+            }
+        }
+        let sha3_digest = hasher.finalize().digest;
+        if *expected_sha3_digest != sha3_digest {
+            return Err(SuiError::from("Sha does not match"));
+        }
+        batch.write()?;
+        Ok(())
```

Comment:
Doing a bit of bookkeeping here to make sure we're on the same page. There are a few other tables that will need to be restored at some point I think.

`root_state_hash_by_epoch`: This can be restored during the orchestration piece I'm going to write by accumulating the live object set up to this point. We only need to write the last root state hash, not all of them, for things to work.

`epoch_start_configuration`: This will get written when we open the new perp store.

`pruned_checkpoint`: should we be technically populating this given that the formal snapshot restore initial state is equivalent to an aggressive prune? Or will it not matter?

transactions and effects: will be part of the archival storage rather than the snapshot. (However I will need to confirm that the protocol doesn't have any lingering dependencies on these when restoring from a formal snapshot with a wiped db)

`executed_transactions_to_checkpoint`: @lxfind can you comment on what the side effects of starting from a snapshot with this table empty will be? And what, if anything, needs to be done. Technically all transactions before this point have been finalized as we are at epoch boundary, but we actually don't have any transactions saved or executed locally. We will be adding some special logic to checkpoint executor so that it does not start from checkpoint 0. Is this all fine w.r.t finalized transactions?

`expected_network_sui_amount` and `expected_storage_fund_imbalance`: these two I think we will need to fix (probably in orchestration PR, not here). If we restore from a snapshot, and we assume that the network doesn't have any inconsistency (given that we would otherwise know from paranoid checks), can we regenerate these values from the live object set? And trust that that value is correct? Perhaps the safest and most efficient thing would be for us to simply add this to the snapshot, that way it is network specific, but then we would need to have a story around verification. Thoughts @lxfind ?

cc @mystenmark as we were discussing this yesterday.

Reply:
@williampsmith I've added the code to restore these tables:
`root_state_hash_by_epoch`, `epoch_start_configuration` and `pruned_checkpoint` from the list above. Rest, we can add as and when we get more clarity from other folks or during the restore orchestration like you mentioned.

Reply:
I think we can trust the snapshot.